src/WebAuthn/Packed.hs

```diff
@@ -34,14 +45,49 @@ verify :: Stmt
   -> ByteString
   -> Digest SHA256
   -> Either VerificationFailure ()
-verify (Stmt _ sig cert) ad adRaw clientDataHash = do
+verify (Stmt algo sig cert) ad adRaw clientDataHash = do
   let dat = adRaw <> BA.convert clientDataHash
   case cert of
     Just x509 -> do
-      let pub = X509.certPubKey $ X509.getCertificate x509
-      verifyX509Sig (X509.SignatureALG X509.HashSHA256 X509.PubKeyALG_EC) pub dat sig "Packed"
+        let x509Cert = X509.getCertificate x509 
+            pub = X509.certPubKey x509Cert
+        verifyX509Sig (X509.SignatureALG X509.HashSHA256 X509.PubKeyALG_EC) pub dat sig "Packed"
+        certMeetsCriteria x509Cert
     Nothing -> do
       pub <- case attestedCredentialData ad of
           Nothing -> Left MalformedAuthenticatorData
           Just c -> parsePublicKey $ credentialPublicKey c
+      when (not $ hasMatchingAlg pub algo) $ Left MalformedAuthenticatorData
       verifySig pub sig dat
+    where
+        certMeetsCriteria :: X509.Certificate -> Either VerificationFailure ()
+        certMeetsCriteria c = do
+            let maaguid = unAAGUID . aaguid <$> attestedCredentialData ad
+                (X509.Extensions mX509Exts) = X509.certExtensions c
+                mX509Ext = mX509Exts >>= findProperExtension [1,3,6,1,4,1,45724,1,1,4]
+                dnElements = X509.getDistinguishedElements $ X509.certSubjectDN c
+            adAAGUID <- maybe (Left $ MalformedX509Certificate "No AAGUID provided in attested credential data") (return . unAAGUID . aaguid) $ attestedCredentialData ad
+            certAAGUID <- maybe (Left $ MalformedX509Certificate "No AAGUID present in x509 extensions") (decodeAAGUID . X509.extRawContent) mX509Ext
+            when (certAAGUID /= adAAGUID) . Left . MalformedX509Certificate $ "AAGUID in attested credential data doesn't match the one in x509 extensions"
+            when (not 
+                (hasDnElement X509.DnCountry dnElements)
+                &&
+                (hasDnElement X509.DnOrganization dnElements)
+                &&
+                (hasDnElement X509.DnCommonName dnElements)
+                &&
+                (findDnElement X509.DnOrganizationUnit dnElements == Just "Authenticator Attestation")) . Left $ MalformedX509Certificate "Certificate SubjectDN doesn't meet crtieria"
+        hasDnElement :: X509.DnElement -> [(OID.OID, X509.ASN1CharacterString)] -> Bool
+        hasDnElement el = isJust . findDnElement el
+        findDnElement :: X509.DnElement -> [(OID.OID, X509.ASN1CharacterString)] -> Maybe X509.ASN1CharacterString
+        findDnElement dnElementName = fmap snd . find ((==) (OID.getObjectID dnElementName) . fst)
+        findProperExtension :: OID.OID -> [X509.ExtensionRaw] -> Maybe X509.ExtensionRaw
+        findProperExtension extensionOID = find ((==) extensionOID . X509.extRawOID)
+        showByteString :: Maybe ByteString -> Text
+        showByteString bs = fromMaybe "" (pack . unpack <$> bs)
```

Comment:
Importing ByteString's unpack and Text's pack unqualified is not great.

Reply:
well - to be more precise - this is a function that is not in use anymore (I was actually using it while troubleshooting the AAGUID case). Removing the whole thing :)